docs/prover/approx/grounding.md

```diff
@@ -0,0 +1,274 @@
+
+In most contexts, you may not have a `forall` expression contained inside of an
+`exists` statement.  For example, this is allowed:
+
+```cvl
+assert forall address x . forall uint y . exists uint z . exists uint w . e(x,y,z,w);
+```
+
+but this is disallowed:
+
+```cvl
+assert forall address x . exists address y . exists address z . forall address w . e(x,y,z,w);
+```
+
+In the latter case, the `forall address w` is contained inside the `exists address z`.
+
+Logical negations and `require` statements reverse the rules for `forall` and
+`exists` statements: in those contexts, you cannot nest an `exists` expression
+inside of a `forall` statement.  Including another negation will again reverse
+the rules.  For example, the following are allowed:
+
+```cvl
+require    exists address x . forall uint y . e(x,y);
+assert   !(exists address x . forall uint y . e(x,y));
+require  !(forall address x . exists uint y . e(x,y));
+assert !(!(forall address x . exists uint y . e(x,y)));
+```
+
+but these are disallowed:
+
+```cvl
+require  forall address x . exists uint y . e(x,y);
+assert !(forall address x . exists uint y . e(x,y));
+```
+
+One common way to work around these limitations is by replacing `exists`
+expressions with concrete expressions that produce the values that should exist.
+
+For example, suppose your contract function always returned twice its input:
+
+```solidity
+function f(uint x) external returns(uint) { return 2 * x; }
+```
+
+You might like to prove that if `x` is positive, then it's possible for `f` to
+output something smaller than `f(x)`:
+
+```cvl
+assert forall uint x . (x > 0) => (exists uint y . f(y) < f(x));
+```
+
+This is clearly true; for example `f(x - 1)` is always smaller than `f(x)`, as
+is `f(0)`.  However, to work around the nested quantifier restriction, we have
+to help the Prover find the correct value for `y`.  We could replace this
+statement with either of the following two:
+
+```cvl
+assert forall uint x . (x > 0) => f(0) < f(x);
+assert forall uint x . (x > 0) => f(x-1) < f(x);
+```
+
+(grounding-recursion)=
+### Recursion
+
+Quantified statements that relate a function with itself on two different
+inputs may give incorrect counterexamples.  For example, the following `forall`
+statement refers to `f` twice:
+
+```
+rule recursiveQuantifier {
+    require forall uint x . f(x) > f(x-1);
+    assert f(8) > f(6);
+}
+```
+
+Although we can see that the assertion must be true, we would need to combine
+the statements `f(8) > f(7)` and `f(7) > f(6)` to prove it, and the grounding
+mechanism is unable to do this.
+
+In these cases, you may see a counterexample that doesn't satisfy the `require`
+statement.
+
+```{todo}
+Do we have general suggestions for rewriting these?
+```
+
+(grounding-arguments)=
+### Variables must be arguments
+
+In order for grounding to work, every variable appearing in a quantified
+statement must be used at least once as an argument to a contract function.
```

Comment:
"contract function" -> "function"
Actually more correct would be "ghost function", but I think we do sort of all contract function calls, though these are very often buggy.
You can write "contract/ghost function", but we should say something somewhere saying that calling contract functions inside quantified expressions is dubious.